hclsyntax/parser.go

```diff
@@ -1161,15 +1161,19 @@ func (p *parser) finishParsingFunctionCall(name Token) (Expression, hcl.Diagnost
 	for openTok.Type == TokenDoubleColon {
 		nextName := p.Read()
 		if nextName.Type != TokenIdent {
-			diags = append(diags, &hcl.Diagnostic{
+			diag := hcl.Diagnostic{
 				Severity: hcl.DiagError,
 				Summary:  "Missing function name",
 				Detail:   "Function scope resolution symbol :: must be followed by a function name in this scope.",
 				Subject:  &nextName.Range,
 				Context:  hcl.RangeBetween(name.Range, nextName.Range).Ptr(),
-			})
+			}
+			diags = append(diags, &diag)
 			p.recoverOver(TokenOParen)
-			return nil, diags
+			return &ExprSyntaxError{
+				ParseDiags:  hcl.Diagnostics{&diag},
+				Placeholder: cty.DynamicVal,
```

Comment:
I think `cty.DynamicVal` is helpful because it means that the expression result will still be something usable in expressions, even though it'll probably just produce another unknown value as the result. If we use `cty.NilVal` then we'll inevitably end up having to add nil value checks everywhere to avoid panicking when evaluating expressions containing syntax errors.

Reply:
Thanks for explaining, @apparentlymart 👍